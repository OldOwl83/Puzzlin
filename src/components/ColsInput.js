import { useContext, useState, useEffect } from "react";
import { PuzzleGrid } from "../RompeCocosApp";

export const ColsInput = () => {

    const { puzzleGrid, setPuzzleGrid } = useContext( PuzzleGrid );
    
    const [ input, setInput ] = useState( puzzleGrid[0] );
    
    useEffect(() => {
        
        if( input >= 3 || input == '' )
        {
            if( input >= 3 )
                setPuzzleGrid( [ Number( input ), puzzleGrid[1] ] );
            else
                setPuzzleGrid( [ Number( 3 ), puzzleGrid[1] ] );
        }else
        {
            setPuzzleGrid( [ Number( 3 ), puzzleGrid[1] ] );
            setInput( 3 );
        }
    }, [ input, setPuzzleGrid ]);

    return <input 
                type="number" 
                value={ input }
                min="3"
                onChange={ ( e ) => setInput( e.target.value) } 
            />
}
