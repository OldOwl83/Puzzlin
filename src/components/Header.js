import React from 'react';
import { Link, useLinkClickHandler } from 'react-router-dom';

export const Header = () => {

    const linker = useLinkClickHandler( '/', { 
        target: '_self',
        replace: false,
        state: {
            contScroll: 0,
            selectedPhoto: 'images/portada1.png',
            author: 'Mauro Donnantuoni',
        }
    });

    return (
        <>
        <Link to='/' id='toHomeLink' onClick={ linker }>
            <header>
                <h1>Puzzlin</h1>
                <div id="logoContainer">
                    <img src="images/rompecabezas.png" alt="logo" />
                </div>
            </header>
        </Link>

        <hr />
        </>
    );
};
