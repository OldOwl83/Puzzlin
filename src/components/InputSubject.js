import { useState } from "react";
import { PropTypes } from 'prop-types';
import { useNavigate } from "react-router-dom";

import '../styles/InputSubject.css';

export const InputSubject = ( { setSearch } ) => {

    // console.log( 'InputSubject' );

    const navigate = useNavigate();

    const [ input, setInput ] = useState( '' );
    
    const submitInput = (e) => {
        e.preventDefault();
        
        setInput( '' );

        setSearch( input );

        navigate( `../${ input }` );
    }

    return (

        <h2>Input a subject and select a photo. Then, press on the preview to start.
            <form onSubmit={ submitInput }>
                <input 
                    id="inputSubject"
                    type="text"
                    autoComplete="on"
                    value={ input }
                    onChange={ (e) => setInput( e.target.value ) }
                />
            </form>
        </h2>
    )
}

InputSubject.propTypes = {

    setSearch: PropTypes.func.isRequired,
}