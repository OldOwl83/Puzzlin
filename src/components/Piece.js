import { useContext, useEffect, useRef } from 'react';
// import PropTypes from 'prop-types';

import '../styles/Piece.css';
import { MainPhotoContext, PuzzleGrid } from '../RompeCocosApp';
// import { columns } from './Puzzle';

export const Piece = ( { position, direc } ) => {

    const { photoPuzzle } = useContext( MainPhotoContext );

    const { puzzleGrid } = useContext( PuzzleGrid );

    const previousPosition = useRef( [] );

    useEffect( () => {
       
    previousPosition.current = position;
    
    }, [ position, direc ]);

    return (
        <div className={ position === 'black' ? `blackPieceContainer black${ direc[0] }` : `pieceContainer ${ previousPosition.current !== position && position !== 'black' ? direc[0] : '' }` }>
            <img 

                className={ ( position === 'black' && 'imgBlack' ) || 'imgPiece'}
                src={ photoPuzzle }
                style={ position !== 'black' 
                    ? { left: `${position[0] * -100}%`, top: `${position[1] * -100}%`, width: `${ puzzleGrid[0] * 100}%` } 
                    : {}}

                alt="Piece of puzzle"
            />
        </div>
    )
}

// Piece.PropTypes = {
//     photo: PropTypes.string.isRequired,
//     position: PropTypes.array[Any(Number), Any(Number)].isRequired
// }
