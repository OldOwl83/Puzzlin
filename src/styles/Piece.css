.pieceContainer{
    position: relative;
    
    width: 98%;
    height: 98%;
    overflow: hidden;
    
    border: 1px solid gray;
    border-radius: 8px;
    box-sizing: border-box;
}

.blackPieceContainer{
    display: flex;
    justify-content: center;
    align-items: center;
    
    width: 98%;
    height: 98%;
    
    border: 2px solid black;
    border-radius: 8px;
    background: radial-gradient(beige 0%, #333 90%);
    box-sizing: border-box;
}

.imgPiece{
    position: absolute;
}

.imgBlack{
   
    width: 80%;
    height: 70%;

    border-radius: 8px;
}

.up{
    animation: goUp 0.7s;
}

.down{
    animation: goDown 0.7s;
}

.left{
    animation: goLeft 0.7s;
}

.right{
    animation: goRight 0.7s;
}

@keyframes goUp{
    from { transform: translateY(100%); }
    to { transform: translateY(0%);}
}

@keyframes goDown{
    from { transform: translateY(-100%); }
    to { transform: translateY(0%);}
}

@keyframes goLeft{
    from { transform: translateX(100%); }
    to { transform: translateY(0%);}
}

@keyframes goRight{
    from { transform: translateX(-100%); }
    to { transform: translateY(0%);}
}

/* It must redefined the same animation for the black piece, to force to take it when two piece exchange positions twice or more consecutively. */
.blackdown{
    animation: blackGoUp 0.7s;
}

.blackup{
    animation: blackGoDown 0.7s;
}

.blackright{
    animation: blackGoLeft 0.7s;
}

.blackleft{
    animation: blackGoRight 0.7s;
}

@keyframes blackGoUp{
    from { transform: translateY(100%); }
    to { transform: translateY(0%);}
}

@keyframes blackGoDown{
    from { transform: translateY(-100%); }
    to { transform: translateY(0%);}
}

@keyframes blackGoLeft{
    from { transform: translateX(100%); }
    to { transform: translateY(0%);}
}

@keyframes blackGoRight{
    from { transform: translateX(-100%); }
    to { transform: translateY(0%);}
}
