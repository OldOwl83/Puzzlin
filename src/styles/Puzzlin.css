h1{
    color: rgb(1, 80, 83);
    font-size: 2.4rem;
    padding: 0 40px;
}

hr{
    border-color: beige;
}