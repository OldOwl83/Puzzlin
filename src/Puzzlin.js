import { 
    BrowserRouter as Router,
    Routes, Route,
        } from "react-router-dom"; 

import { createContext, useState } from "react";
import { PhotoSelector } from "./components/PhotoSelector";
import { Puzzle } from "./components/Puzzle";
import { NoRoute } from "./components/NoRoute";

import './styles/Puzzlin.css';

export const MainPhotoContext = createContext( null );
export const PuzzleGrid = createContext( [] );

export const Puzzlin = () => {

    const [ puzzlePhoto, setPuzzlePhoto ] = useState( '' );
    const [ puzzleGrid, setPuzzleGrid ] = useState( [4, 3] );

    return (
        <>
            <h1>Puzzlin</h1>
            <hr />

            <Router>
                <MainPhotoContext.Provider value={ { puzzlePhoto, setPuzzlePhoto } } >
                <PuzzleGrid.Provider value={ { puzzleGrid, setPuzzleGrid } }>

                    <Routes>

                        <Route path="/" element={ <PhotoSelector /> } />
                        <Route path="/?q=:query" element={ <PhotoSelector /> } />

                        <Route path="/puzzle" element={ <Puzzle /> } />

                        <Route path="/*" element={ <NoRoute /> } />

                    </Routes>

                </PuzzleGrid.Provider>
                </MainPhotoContext.Provider>
            </Router>

        </>
    )
};