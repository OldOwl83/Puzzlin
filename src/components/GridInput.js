/*
* It handles the contextual values of the puzzle grid. These values represent the columns and rows to cut out the puzzle photo and are in an array with two indices, which is a state belonging to the <Puzzlin /> component.
*/

import { useContext, useState, useEffect } from "react";

import { PuzzleGrid } from "../Puzzlin";
import { NumberInput } from "./NumberInput";

import "../styles/GridInput.css";

export const GridInput = () => {

    const { puzzleGrid, setPuzzleGrid } = useContext( PuzzleGrid );
    
    const [ cols, setCols ] = useState( puzzleGrid[0] );
    const [ rows, setRows ] = useState( puzzleGrid[1] );
    
    useEffect(() => {
        
        setPuzzleGrid( [ cols, rows ] );
        
    }, [ setPuzzleGrid, cols, rows ]);

    return (
    
        <div id="GridInput">
            <p>Columns
                <NumberInput setValue={ setCols } min={ 4 } max={ 8 } initialValue={ puzzleGrid[0] } />
                {/* It send de previous context values to initial value of this component, to maintain them across all routes. */}
            </p>
            <p>Rows
                <NumberInput setValue={ setRows } min={ 3 } max={ 8 } initialValue={ puzzleGrid[1] } />
            </p>
        </div>
    );
}
