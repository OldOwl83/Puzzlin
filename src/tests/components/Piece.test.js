import 'react';
import { mount } from 'enzyme';

import { Piece } from '../../components/Piece';
import { MainPhotoContext, PuzzleGrid } from '../../Puzzlin';


jest.mock( 'react', () => ({
    ...jest.requireActual( 'react' ),
    useRef: () => { return { current: [ 1, 2 ] }; },
 }));

describe('Tests on "white" <Piece /> component', () => {

    let position = [ 1, 1 ];
    let puzzleGrid = [ 4, 3 ];

    const wrapper = mount(
        <MainPhotoContext.Provider value={ { puzzlePhoto: 'portada.jpg' } }>
        <PuzzleGrid.Provider value={ { puzzleGrid: puzzleGrid } } >

            <Piece position={ position } direc={ 'up' } />

        </PuzzleGrid.Provider>
        </MainPhotoContext.Provider>
    );
  

    test('It should match the snapshot', () => {
      
        expect( wrapper ).toMatchSnapshot();
    });

    test('The image should take the source passed by parameter.', () => {
      
        expect( wrapper.find( 'img' ).prop( 'src' )).toBe( 'portada.jpg' );
    });

    test('The "div" and "img" classNames should be "pieceContainer up" and "imgPiece", respectively.', () => {
      
        expect( wrapper.find( 'div' ).prop( 'className' )).toBe( 'pieceContainer up' );
        expect( wrapper.find( 'img' ).prop( 'className' )).toBe( 'imgPiece' );
    });

    test('It should sets the right position and width of photo.', () => {
      
        expect( wrapper.find( 'img' ).prop( 'style' ).left).toBe( `${ position[0] * -100 }%` );
        expect( wrapper.find( 'img' ).prop( 'style' ).top).toBe( `${ position[1] * -100 }%` );
        expect( wrapper.find( 'img' ).prop( 'style' ).width).toBe( `${ puzzleGrid[0] * 100 }%` );
    });
});

describe('Tests on "black" <Piece /> component', () => {

    let position = 'black';
    let puzzleGrid = [ 4, 3 ];

    const wrapper = mount(
        <MainPhotoContext.Provider value={ { puzzlePhoto: 'portada.jpg' } }>
        <PuzzleGrid.Provider value={ { puzzleGrid: puzzleGrid } } >

            <Piece position={ position } direc={ 'up' } />

        </PuzzleGrid.Provider>
        </MainPhotoContext.Provider>
    );
  

    test('It should match the snapshot', () => {
      
        expect( wrapper ).toMatchSnapshot();
    });

    test('The image should take the source passed by parameter.', () => {
      
        expect( wrapper.find( 'img' ).prop( 'src' )).toBe( 'portada.jpg' );
    });

    test('The "div" and "img" classNames should be "blackPieceContainer blackup" and "imgBlack", respectively.', () => {
      
        expect( wrapper.find( 'div' ).prop( 'className' )).toBe( 'blackPieceContainer blackup' );
        expect( wrapper.find( 'img' ).prop( 'className' )).toBe( 'imgBlack' );
    });
});
