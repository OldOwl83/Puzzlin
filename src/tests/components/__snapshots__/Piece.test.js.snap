// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on "black" <Piece /> component It should match the snapshot 1`] = `
<div
  className="blackPieceContainer blackup"
>
  <img
    alt="Piece of puzzle"
    className="imgBlack"
    draggable="false"
    src="portada.jpg"
    style={Object {}}
  />
</div>
`;

exports[`Tests on "white" <Piece /> component It should match the snapshot 1`] = `
<div
  className="pieceContainer up"
>
  <img
    alt="Piece of puzzle"
    className="imgPiece"
    draggable="false"
    src="portada.jpg"
    style={
      Object {
        "left": "-100%",
        "top": "-100%",
        "width": "400%",
      }
    }
  />
</div>
`;
